### PR TITLE
[data] Refactor planner.py

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

`planner.py` has too many if-else branches. refactor code to make the code cleaner and easier to extend in the future.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/planner/planner.py [hunk 1]
```diff
@@ -1,28 +1,98 @@
...
+    def plan_input_data_op(
+        logical_op: InputData, physical_children: List[PhysicalOperator]
+    ) -> PhysicalOperator:
+        """Get the corresponding DAG of physical operators for InputData."""
+        assert len(physical_children) == 0
+
+        return InputDataBuffer(
+            input_data=logical_op.input_data,
+            input_data_factory=logical_op.input_data_factory,
+        )
```
Comment (bveeramani) [inline python/ray/data/_internal/planner/planner.py:61]: Nit: IMO this'll be easier to read if we consistently define plan functions in separate modules. I think mixing the two approaches (function definitions in separate modules and in this function) hurts readability.

Not heavily opinionated on this, so am okay to keep as is.

File: python/ray/data/_internal/planner/planner.py [hunk 2]
```diff
@@ -1,28 +1,98 @@
...
+@DeveloperAPI
+def register_plan_logical_op_fn(
+    logical_op_type: Type[LogicalOperator],
+    plan_fn: PlanLogicalOpFn,
+):
+    """Register a plan function for a logical operator type."""
+    PLAN_LOGICAL_OP_FNS.append((logical_op_type, plan_fn))
+
+
+def register_plan_logical_op_fns():
```
Comment (bveeramani) [inline python/ray/data/_internal/planner/planner.py:30]: Nit: Maybe something like this to disambiguate from `register_plan_logical_op_fn`? 
```suggestion
def register_default_plan_logical_op_fns():
```

From the name, I'd expect something like this:
```
register_plan_logical_op_fns(plan_fns: List[PlanLogicalOpFn])
```

File: python/ray/data/_internal/planner/planner.py [hunk 3]
```diff
@@ -1,28 +1,98 @@
...
-from ray.data._internal.planner.plan_all_to_all_op import plan_all_to_all_op
-from ray.data._internal.planner.plan_from_op import plan_from_op
-from ray.data._internal.planner.plan_input_data_op import plan_input_data_op
-from ray.data._internal.planner.plan_limit_op import plan_limit_op
-from ray.data._internal.planner.plan_read_op import plan_read_op
-from ray.data._internal.planner.plan_udf_map_op import plan_udf_map_op
-from ray.data._internal.planner.plan_write_op import plan_write_op
+from ray.util.annotations import DeveloperAPI
+
+LogicalOperatorType = TypeVar("LogicalOperatorType", bound=LogicalOperator)
```
Comment (bveeramani) [inline python/ray/data/_internal/planner/planner.py:11]: What're the advantages of using this over `Type[LogicalOperator]`?

**Review by bveeramani — APPROVED** [review]

File: python/ray/data/_internal/planner/planner.py (hunk 1 above)
Comment (raulchen) [reply python/ray/data/_internal/planner/planner.py:61]: some plan functions are just one line. I feel it's simpler and cleaner to just put them here. For those complex plan functions, it's better to put them in separate files.

File: python/ray/data/_internal/planner/planner.py (hunk 2 above)
Comment (raulchen) [reply python/ray/data/_internal/planner/planner.py:30]: makes sense

File: python/ray/data/_internal/planner/planner.py (hunk 3 above)
Comment (raulchen) [reply python/ray/data/_internal/planner/planner.py:11]: The function argument should be a subclass  of LogicalOperator. If using `Type[LogicalOperator]`, it requires exactly `LogicalOperator`.

**Pull request merged**